addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java

```diff
@@ -0,0 +1,188 @@
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.harmonyhub.HarmonyHubBindingConstants;
+import org.openhab.binding.harmonyhub.config.HarmonyDeviceConfig;
+import org.openhab.binding.harmonyhub.internal.HarmonyHubHandlerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.whistlingfish.harmony.config.ControlGroup;
+import net.whistlingfish.harmony.config.Device;
+import net.whistlingfish.harmony.config.Function;
+
+/**
+ * The {@link HarmonyDeviceHandler} is responsible for handling commands for Harmony Devices, which are
+ * sent to one of the channels. It also is responsible for dynamically creating the button press channel
+ * based on the devices available button press functions.
+ *
+ * @author Dan Cunningham - Initial contribution
+ *
+ */
+public class HarmonyDeviceHandler extends BaseThingHandler implements HubStatusListener {
+
+    private Logger logger = LoggerFactory.getLogger(HarmonyDeviceHandler.class);
+
+    public final static Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(HARMONY_DEVICE_THING_TYPE);
+
+    HarmonyHubHandler bridge;
+    HarmonyHubHandlerFactory factory;
+    int id;
+    String name;
+    String logName;
+
+    public HarmonyDeviceHandler(Thing thing, HarmonyHubHandlerFactory factory) {
+        super(thing);
+        this.factory = factory;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.trace("Command {}  for {}", command, channelUID);
+        Channel channel = getThing().getChannel(channelUID.getId());
+
+        if (channel == null) {
+            logger.warn("No such channel {] for device {}", channelUID, getThing());
+            return;
+        }
+
+        if (!(command instanceof StringType)) {
+            logger.warn("Command {} is not a String type for channel {] for device {}", command, channelUID,
+                    getThing());
+            return;
+        }
+
+        logger.debug("Pressing button {} on {}", command, id > 0 ? 0 : name);
+
+        if (id > 0) {
+            bridge.getClient().pressButton(id, command.toString());
+        } else {
+            bridge.getClient().pressButton(name, command.toString());
+        }
+
+        // may need to ask the list if this can be set here?
+        updateState(channelUID, UnDefType.UNDEF);
+    }
+
+    @Override
+    public void initialize() {
+        id = getConfig().as(HarmonyDeviceConfig.class).id;
+        name = getConfig().as(HarmonyDeviceConfig.class).name;
+
+        if (name == null && id <= 0) {
+            logger.error("A harmony device thing must be configured with a device name OR a postive device id");
+            updateStatus(ThingStatus.UNINITIALIZED, ThingStatusDetail.CONFIGURATION_ERROR);
+        } else {
+            logName = id > 0 ? String.valueOf(id) : name;
+            logger.debug("initializing {}", logName);
+            updateStatus(ThingStatus.INITIALIZING);
```

Comment:
When testing, for some reason all my devices stayed in status INITIALIZING and never went ONLINE.